src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,78 @@
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.VisualStudio.Text.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.RenameTracking
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, LanguageNames.VisualBasic,
+        Name = nameof(RenameTrackingCodeRefactoringProvider)), Shared]
+    internal class RenameTrackingCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        public const string DiagnosticId = "RenameTracking";
+        public static DiagnosticDescriptor DiagnosticDescriptor = new DiagnosticDescriptor(
+            DiagnosticId, title: "", messageFormat: "", category: "",
+            defaultSeverity: DiagnosticSeverity.Hidden, isEnabledByDefault: true,
+            customTags: DiagnosticCustomTags.Microsoft.Append(WellKnownDiagnosticTags.NotConfigurable));
+
+        public const string RenameFromPropertyKey = "RenameFrom";
+        public const string RenameToPropertyKey = "RenameTo";
+
+        private readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
+        private readonly IEnumerable<IRefactorNotifyService> _refactorNotifyServices;
+
+        [ImportingConstructor]
+        public RenameTrackingCodeRefactoringProvider(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            [ImportMany] IEnumerable<IRefactorNotifyService> refactorNotifyServices)
+        {
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _refactorNotifyServices = refactorNotifyServices;
+        }
+
+        // Internal for testing purposes
+        internal async Task<Diagnostic?> TryGetDiagnosticAsync(Document document, CancellationToken cancellationToken)
+        {
+            var syntaxTree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            return RenameTrackingTaggerProvider.TryGetDiagnostic(
+                syntaxTree, DiagnosticDescriptor, cancellationToken);
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, span, cancellationToken) = context;
+
+            // Ensure rename can still be invoked in this document. We reanalyze the document for
+            // diagnostics when rename tracking is manually dismissed, but the existence of our
+            // diagnostic may still be cached, so we have to double check before actually providing
+            // any fixes.
+            if (!RenameTrackingTaggerProvider.CanInvokeRename(document))
+                return;
+
+            var diagnostic = await TryGetDiagnosticAsync(document, cancellationToken).ConfigureAwait(false);
+            if (diagnostic == null)
+                return;
```

Comment:
note: the impl still uses the existing codepaths to create diagnostics and then create code actions based on them.  Note though that the diagnostics are never reported by a diagnostic analyzer.  They're just an internal impl detail that i'm keeping to keep this PR small.  I will have a followup PR that removes this intermediary unnecessary translation from state->diagnostic->action.